### ISSUE TITLE
(35) Insufficent Randomness (for Windows 7 with OpenSSL 3.0.3)

### ISSUE THREAD

**Issue by DanTheMann15** [curl/curl#8849]
Since curl was updated with OpenSSL 3.0.3, i started getting error 35 when trying to get info on ipsw files hosted on apple's servers.
i'm not exactly sure why this issue is occuring, but here's some information that should help us track it down a bit.

OS: Windows 7 SP1 (x86_64)
Version(s): CURL 7.83.1 (7.83.0 with OpenSSL 3.0.3)
Downloaded From: https://curl.se/windows/

-V Output: 
> curl 7.83.1 (x86_64-pc-win32) libcurl/7.83.1 OpenSSL/3.0.3 (Schannel) zlib/1.2.12 brotli/1.0.9 libidn2/2.3.2 libssh2/1.10.0 nghttp2/1.47.0 libgsasl/1.10.0
> Release-Date: 2022-05-11
> Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
> Features: alt-svc AsynchDNS brotli gsasl HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI TLS-SRP UnixSockets

How to Reproduce (example):
curl -i https://updates.cdn-apple.com/2022FCSWinter/fullrestores/002-81308/6BD01782-B7EA-4BF7-987A-C3D3F5755CC0/iPhone_4.0_64bit_15.4.1_19E258_Restore.ipsw

Expected (Something like this):

> HTTP/1.1 200 OK
> Content-Type: binary/octet-stream
> Content-Length: 5216804428
> Last-Modified: Mon, 28 Mar 2022 01:13:50 GMT
> x-amz-meta-digest-sh1: bcb98612c204172a1b1a7c12577ffcdf38e0ead0
> x-amz-meta-digest-sha256: 73a5eb08698c2ce1eb6499824800c09c855ad1cb53e7f8028c4410ee72599b93
> x-amz-version-id: 2J8Pjdc3WOpauaC40VRcwcTm.9YLm1Cb
> Accept-Ranges: bytes
> Server: AmazonS3
> Date: Sat, 14 May 2022 12:20:02 GMT
> X-Cache: RefreshHit from cloudfront, hit-fresh, miss, none
> Via: 1.1 b55214c298a6470ca4f296a9e493d5f2.cloudfront.net (CloudFront), http/1.1 uschi5-edge-lx-003.ts.apple.com (acdn/14
> 4.13255), http/1.1 uschi5-edge-bx-017.ts.apple.com (acdn/144.13255), http/1.1 uschi5-edge-bx-017.ts.apple.com (acdn/144.
> 13255)
> X-Amz-Cf-Pop: ORD52-C1
> X-Amz-Cf-Id: M5R-liZCWATc_FUuZIkMJX4_Uog2gHIFf4GaCkvSUb-nzaNNKyGV6g==
> Cache-Control: max-age=21600, public
> Etag: "b9ea458237a909061b66d2a7f7477b61-622"
> Age: 14206
> CDNUUID: aba81b1c-2462-4dba-9587-32816591578e-614671188
> Connection: keep-alive
> 
> Warning: Binary output can mess up your terminal. Use "--output -" to tell
> Warning: curl to output it to your terminal anyway, or consider "--output
> Warning: <FILE>" to save to a file.

Actual Result:
> curl: (35) Insufficient randomness

**Comment by bagder** [comment]
My understanding of this is that OpenSSL 3 (actually since OpenSSL version >= 1.1.0) inits the random seed itself, which means that `RAND_status()` *should* return 1. When it doesn't, then something clearly is wrong and libcurl returns error for it:

https://github.com/curl/curl/blob/5227367903597b5053608b599c52644980e925e6/lib/vtls/openssl.c#L483

So... something seems to be wrong with either my assumptions or with the conditions in which the code runs in your system!

**Comment by alexetgus** [comment]
Hi,

I have the same problem with all the URLs. I think I'll do a rollback...

OS : Win7 Pro (x64)
Curl : 7.83.1
OpenSSL : 3.0.3 in Curl, 3.0.1 in Windows system

**Comment by TheXman-TX** [comment]
I am also experiencing the exact same issue as the original poster.  When trying to access **ANY** https site, using the 7.83.1 version of curl.exe on Windows 7 Pro x64, it immediately returns with:

`curl: (35) Insufficient randomness`.  

When I execute the exact same commands with cURL 7.83.0, I get the expected results.  

For the record, when using the 7.83.1 version of curl.exe on Windows 10 Pro x64, it works as expected.


cURL Version Info:
```
curl 7.83.1 (x86_64-pc-win32) libcurl/7.83.1 OpenSSL/3.0.3 (Schannel) zlib/1.2.12 brotli/1.0.9 libidn2/2.3.2 libssh2/1.10.0 nghttp2/1.47.0 libgsasl/1.10.0
Release-Date: 2022-05-11
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli gsasl HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI TLS-SRP UnixSockets
```

**Comment by rsbeckerca** [comment]
Did you build OpenSSL 3.0.3 yourself? If so, what randomizer did you select?

**Comment by TheXman-TX** [comment]
No, I downloaded the latest Windows x64 version directly from the curl.se site on this page:

https://curl.se/windows/

**Comment by TheXman-TX** [comment]
To whom are you asking that question?  If it is to me, the answer is no.  I am using the openssl version that is embedded in the executable.  The executable was downloaded from the curl.se site on the following page:

https://curl.se/windows/

**Comment by rsbeckerca** [comment]
> To whom are you asking that question? If it is to me, the answer is no. I am using the openssl version that is embedded in the executable. The executable was downloaded from the curl.se site on the following page:
> 
> https://curl.se/windows/

Just checking whether a custom build might have chosen a bad randomizer. I have seen that happen when a poorly seeded PRNGD is set up.

**Comment by bagder** [comment]
This sounds like an OpenSSL 3.0.3 (on Windows 7?) bug to me. Do any earlier versions  of OpenSSL 3 work better?

/cc @vszakats

**Comment by DanTheMann15** [comment]
this build of curl (for windows) was the last one to work correctly: https://curl.se/windows/dl-7.83.0/
starting with build `7.83.0_1` when the windows curl builds started being built with openssl 3.0.3.

3.0.2 worked fine, so i am not sure exactly what happened here. it could be a bug in openssl 3.0.3 but it's going to be some time and testing before we figure out the root cause.

**Comment by GaneJ** [comment]
I had the same problem too.
Whether to use .netrc or not.
It didn't happen with version 7.82.0.
It only happen with version 7.83.x

Btw, I'm using Windows 7.

**Comment by rsbeckerca** [comment]
> **Yes, that's certainly it. Version 7.83.1 no longer has a random number generator under Win7...**
 
In our build, we have this:
```
$ curl --version
curl 7.83.1 (nse-tandem-nskJ06) libcurl/7.83.1 OpenSSL/3.0.2 zlib/1.2.8
Release-Date: 2022-05-11
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp smtp smtps telnet tftp
Features: alt-svc HSTS HTTPS-proxy IPv6 libz NTLM NTLM_WB SSL TLS-SRP UnixSockets
$ curl --engine list
Build-time engines:
  dynamic
```
This applies to both ia64, using `PRNGD`, and x86, using `rdrand` from OpenSSL.

**Comment by vszakats** [comment]
Thanks for the tests to all.

If I summarize correctly, both curl [7.83.0](https://curl.se/windows/dl-7.83.0_1/curl-7.83.0_1-win64-mingw.tar.xz) and [7.83.1](https://curl.se/windows/dl-7.83.1/curl-7.83.1-win64-mingw.tar.xz) built with OpenSSL 3.0.3, x64 flavour, when run on Win7, will fail with not enough randomness. [7.83.0 + 3.0.2](https://curl.se/windows/dl-7.83.0/curl-7.83.0-win64-mingw.tar.xz) works fine. Buggy combinations will list _no engines_ with `curl --engine list`, while good ones list `rdrand` and `dynamic`.

The single constant component in the failing cases is OpenSSL 3.0.3. The way OpenSSL is built did not change between 3.0.2 and 3.0.3. This indicates it either to be a 3.0.3 bug, or an update that required some change in the way OpenSSL is built. `{NEWS,CHANGES}.md` does not suggest anything like that, and nothing related to random generators or engines.

[List of changes](https://github.com/openssl/openssl/compare/openssl-3.0.2...openssl-3.0.3) between 3.0.3 and 3.0.2. 90 commits in total.

One suspicious and deep change is their full switch to `OPENSSL_strcasecmp()` from `strcasecmp()`. In curl-for-win builds, the function translates to `stricmp_l()`. This is locale-dependent comparison, that also requires explicit locale initialization, which is supposed to be done by OpenSSL automatically. It's possible (likely?) that the new, locale-dependent string comparator fails depending on the environment.

Probably the best would be report this to OpenSSL.

**Comment by rsbeckerca** [comment]
> Probably the best would be report this to OpenSSL.

I have already reported a problem for `OPENSSL_strcasecmp()` as it did not build on the x86 variant of my platform. We do not have 3.0.3 as a result. However, it does work fine on `ia64` with 3.0.3, but our OpenSSL build explicitly specifies the randomization source.

**Comment by vszakats** [comment]
@rsbeckerca So, building with `--with-rand-seed=os` may fix this particular problem (with the cost of not allowing a custom random generator — likely a non-issue for most). But, I wonder how many other, less obvious issues remain with a broken `stricmp()` function.

**Comment by rsbeckerca** [comment]
> @rsbeckerca So, building OpenSSL with `--with-rand-seed=os` may fix this particular problem (with the cost of not allowing a custom random generator — likely a non-issue for most). But, I wonder how many other, less obvious issues remain with a broken `str[n]icmp()` function.

Sounds about right.

**Comment by rsbeckerca** [comment]
This has been opened as https://github.com/openssl/openssl/issues/18322

**Comment by vszakats** [comment]
Test binaries using latest OpenSSL snapshot are available here: https://ci.appveyor.com/project/curlorg/curl-for-win/builds/43573097/artifacts

OpenSSL team is looking for test result here: https://github.com/openssl/openssl/issues/18322

**Comment by vszakats** [comment]
This issue looks much worse than the uninitialized random generator, which at least fails early and clearly.

While this is sorted out, my question is what would be the best temporary solution for curl-for-win builds?

1. Downgrade to OpenSSL 3.0.2? This will re-introduce the [two vulnerabilities](https://www.openssl.org/news/secadv/20220503.txt), one Medium and one with Low severity.
2. Disable OpenSSL altogether and go with Schannel-only? (This will result in missing TLSv3-support on possibly all, but for sure on non-bleeding edge Windows versions.)
3. Replace OpenSSL with LibreSSL? (also as a longer-term solution).
4. Something else?

I tend to prefer option 1.

**Comment by bagder** [comment]
I think (1) is fairly safe. The three mentioned CVEs:
1. We don't use c_rehash
2. we don't use `OCSP_basic_verify()` with `OCSP_NOCHECKS`
3. RC4-MD5 is not enabled by default anyway

**Comment by vszakats** [comment]
@bagder: Sounds good! Also they seem to have approved non-locale dependent string comparisons now.

Anyway, I went along and did a few things to avoid this scenario in the future, also aligned with pending plans to resolve some long-time OpenSSL woes:
- [Re-added](https://github.com/curl/curl-for-win/commit/b67bb573013624aa31edd898f4c7510e9489b58c) support for the latest LibreSSL version (pending [one libssh2](https://github.com/libssh2/libssh2/pull/700) and [one curl](https://github.com/curl/curl/pull/8884) patch). LibreSSL implemented the mitigation that still needs to be force fed to OpenSSL via a custom patch, and generally seems to have better code hygene and non-proprietary build systems (with some non-fatal issues however). It's also good to have an alternative.
- I have earlier implemented build flavours that use Schannel only (e.g. the "mini"). I'm using such build locally to avoid the bulky OpenSSL build (which also started failing a while ago on macOS, and cannot figure out a fix.)
- [Disabled](https://github.com/curl/curl-for-win/commit/817e4fc2661a9a0f2a97c7f16c87299f1e4b0c10) the "DSO" feature in curl-for-win OpenSSL builds, to avoid any chance of loading malicious engines/modules at runtime, via its potentially world-writable config file. (LibreSSL has this feature by default as far as I can tell.)
- Regarding OpenSSL builds, I changed focus and these builds will now serve curl purposes only, so now it's a non-goal to create a general purpose OpenSSL package. This means that for example OpenSSL DLLs and `openssl.exe` may be dropped now from the package (and the build), as those are not necessary for running curl, or linking curl to an app as a DLL. These can make builds faster and downloads smaller. Also, further unnecessary options can be disabled if they create problems or security issues (thinking of the infamous "APPLINK" feature for example, on x86-only).
